Adds registration functionality

Anarox1111 opened the pull request:
This pull request adds full registration functionality. At most, there should be some kind of indicator for end-users that they can log into their accounts after registrering.

Anarox1111 committed:
Implemented registration functionality.

frigvid commented on app/build.gradle.kts:
```diff
@@ -71,6 +71,7 @@ dependencies {
     implementation(libs.navigation.compose)
     implementation(libs.androidx.security.crypto)
     implementation(libs.kotlinx.serialization.json)
+    implementation(libs.androidx.lifecycle.viewmodel.ktx)
```
Why has a dependency been added? I cannot see it being used, nor a reason for why it should be included in the project.

Anarox1111 replied:
I got an unresolved reference `Unresolved reference: viewModelScope`. The only solution I found for this was adding the dependency to `build.gradle` : 

```
dependencies {
    implementation "androidx.lifecycle:lifecycle-viewmodel-ktx:2.5.1"
}
```

Might have stumbled over wrong implementations, at least it fixed my unresolved issue x)

frigvid commented on app/src/main/java/no/usn/mob3000/domain/viewmodel/AuthenticationViewModel.kt:
```diff
@@ -12,17 +12,19 @@ import no.usn.mob3000.domain.model.AuthError
  * @property loginUseCase The Android Use Case handling login business logic.
- * @author frigvid
+ * @author frigvid, Anarox
  * @created 2024-10-21
```
Registration is unique enough that it should have its own view model, e.g. `RegistrationViewModel`.

frigvid commented on app/src/main/java/no/usn/mob3000/domain/viewmodel/AuthenticationViewModel.kt:
```diff
@@ -90,8 +92,39 @@ class LoginViewModel(
+                    val authError = when (error) {
+                        is AuthRestException -> AuthError.fromException(error)
+                        else -> AuthError.Unknown(error.message ?: "Unknown error occurred")
+                    }
```
`AuthError` is used and designed for login's use-case. You are unlikely to get a usable error message, if any, when trying to register an account. If there are multiple different registration event responses like there are for authentication, then registration should have its own.

Anarox1111 replied:
Will look into that.

frigvid commented on app/src/main/java/no/usn/mob3000/ui/screens/auth/CreateUserScreen.kt:
```diff
@@ -85,7 +85,7 @@ fun CreateUserScreen(
 
                 Button(
-                    onClick = { onSignUpClick() },
+                    onClick = { onSignUpClick(mail, password) },
```
Single-use functions do not need brackets. The brackets were only there before in case we needed to do some finagling with additional functions. So this should be `onClick = onSignUpClick(mail, password)`.

frigvid replied:
Also, kind of forgot it in the moment, but you're not doing a password check here. We have two fields and `onSignUpClick` should only be called if both are equal values. If they're not, an error message should display, telling the end-user that both passwords need to be identical.

Anarox1111 replied:
> Single-use functions do not need brackets. The brackets were only there before in case we needed to do some finagling with additional functions. So this should be `onClick = onSignUpClick(mail, password)`.

I know it's not necessary, but I'm using it as a default practice, because it's more practical if changes are applied later on in the future.

frigvid commented on app/src/main/java/no/usn/mob3000/domain/viewmodel/AuthenticationViewModel.kt:
```diff
@@ -90,8 +92,39 @@ class LoginViewModel(
+     * @author Anarox
+     * @created 2024-10-30
+     */
+
```
Remember correct spacing.

frigvid commented on app/src/main/java/no/usn/mob3000/domain/usecase/auth/RegisterUseCase.kt:
```diff
@@ -0,0 +1,27 @@
+*/
+
+
+class RegisterUseCase(private val authRepository: AuthRepository = AuthRepository(AuthDataSource())) {
```
Remember correct spacing. And also remember to document the function. The property needs documentation, and the use case needs documentation. They should not be left blank.

frigvid commented on app/src/main/java/no/usn/mob3000/domain/usecase/auth/RegisterUseCase.kt:
```diff
@@ -0,0 +1,27 @@
+    private fun String.isValidEmail(): Boolean =
+        android.util.Patterns.EMAIL_ADDRESS.matcher(this).matches()
+
+
```
Remember correct spacing.

frigvid commented on app/src/main/java/no/usn/mob3000/ui/screens/auth/CreateUserScreen.kt:
```diff
@@ -32,12 +32,12 @@ import no.usn.mob3000.ui.theme.DefaultButton
  * @created 2024-09-24
  */
 @Composable
 fun CreateUserScreen(
```
Also, since you're using a similar view model function for registration like login, it's important to point out that you're not using it properly. Then you register, the screen receives no return from the view model, and doesn't use it in any way. No loading screen, no error message, and no success message.

In fact, testing it out, the problem of using the `AuthenticationViewModel` becomes apparent. That is designed for `login` specifically, even though `logout` leeches off of it, since it has no special requirements. Any error messages you get, are applied to the `LoginScreen`, which you can see by returning to that screen.

frigvid commented on app/src/main/java/no/usn/mob3000/data/repository/auth/AuthRepository.kt:
```diff
@@ -74,8 +74,25 @@ class AuthRepository(
+
+            currentUserId = authDataSource.getCurrentUserId()
+
+            Result.success(fetchUserData())
```
Had to fix an SMTP issue, but now sign-up actually works again on the back-end. Anyway, that means I got time to test this.

This code will not work, that's why I told you to test what you get back from `signUpWith()`. It doesn't return a session at all, though it does return some data about the user that was created. E.g. the e-mail and so on and so forth. You'll need to rework how registration works. The only way to return the user would be to also call on the signin function inside the register function, but after `signUpWith`, but that does not really tickle my "that sounds like a good idea" bone, so we should find some other way first and compare the two.

Here's an object it returns, with the e-mail stripped.
```
UserInfo(appMetadata={"provider":"email","providers":["email"]}, aud=authenticated, confirmationSentAt=2024-10-31T07:33:06.071847984Z, confirmedAt=null, createdAt=2024-10-31T07:33:06.059134Z, email=[REDACTED], emailConfirmedAt=null, factors=[], id=d6dc8dad-0882-4901-ad54-9738816ad1bb, identities=[Identity(id=d6dc8dad-0882-4901-ad54-9738816ad1bb, identityData={"email":"[REDACTED]","email_verified":false,"phone_verified":false,"sub":"d6dc8dad-0882-4901-ad54-9738816ad1bb"}, identityId=68bd6759-8e9a-4b59-8df9-0402e41fecf5, lastSignInAt=2024-10-31T07:33:06.066021797Z, updatedAt=2024-10-31T07:33:06.06607Z, createdAt=2024-10-31T07:33:06.06607Z, provider=email, userId=d6dc8dad-0882-4901-ad54-9738816ad1bb)], lastSignInAt=null, phone=, role=authenticated, updatedAt=2024-10-31T07:33:07.566847Z, userMetadata={"email":"[REDACTED]","email_verified":false,"phone_verified":false,"sub":"d6dc8dad-0882-4901-ad54-9738816ad1bb"}, phoneChangeSentAt=null, newPhone=null, emailChangeSentAt=null, newEmail=null, invitedAt=null, recoverySentAt=null, phoneConfirmedAt=null, actionLink=null)
```

frigvid reviewed (changes requested):
Changes need to be made before this functionality is ready to be made. See my other comments on the matter.

***e:*** Additionally, remember that pull requests need a description, assignee and a label.

Anarox1111 committed:
Implemented registration functionality.

>>>> Now has unique error handling for registration -> RegistrationError.kt

>>>> error messages applied for strings.xml/ and strings.xml/no

frigvid committed:
feat: Extract Loading and Error progress indicators to components. Extract registration from LoginViewModel. AuthenticationViewModel -> LoginViewModel.

frigvid committed:
ci: Merge remote-tracking branch 'origin/auth_nicco' into auth_nicco

frigvid committed:
chore: Remove unnecessary module dependency.

frigvid committed:
fix: Change result type from registration.

frigvid committed:
feat: Complete registration functionality.

frigvid committed:
chore: Cleanup.

frigvid committed:
fix: Author user name.

Likely more to fix later, but eh.

frigvid committed:
chore: Cleanup.

Truly, to suffer.

frigvid committed:
chore: Remove unnecessary module dependency. Again.

So, apparently, the one time I decided to use the GUI to remove module dependencies is the same time that I find out that it does *not* remove it from `libs.versions.toml`. TIL, I guess.

frigvid reviewed (approved):
Unless I've overlooked something, it should be good now.
